portfolio/src/main/client/src/components/controls/readmore.ts

```diff
@@ -22,34 +28,51 @@ export interface ReadMoreComponent {
  * amount of space.
  *
  * @param content The content to place inside of the readmore.
+ * @param height The height of the readmore when collapsed. Defaults to 20rem.
  * @returns A @see {ReadMoreComponent}
  */
-export const ReadMore = (content: Renderable): ReadMoreComponent => {
+export const ReadMore = (
+  content: Renderable,
+  height = '20rem'
+): ReadMoreComponent => {
   const contentEl: HTMLDivElement =
-    htmlElement`<div class="readmore collapsed">${content}</div>`;
+    htmlElement`<div class="readmore">${content}</div>`;
 
-  const collapse = () => {
-    if (contentEl.classList.contains('collapsed')) return;
+  const component: ReadMoreComponent = {
+    root: contentEl,
+    collapse() {
+      if (this.isCollapsed) return;
+      contentEl.classList.add('collapsed');
 
-    contentEl.classList.add('collapsed');
+      // let CSS transition max height from whatever it was to 0
+      contentEl.style.maxHeight = height;
+      contentEl.style.webkitMaskImage =
+        contentEl.style.maskImage =
+          `linear-gradient(to bottom, 
+                           white calc(${height} - 5rem), 
+                           transparent calc(${height} - 1rem))`;
+      contentEl.dispatchEvent(new Event('readmore-collapse'));
+    },
+    expand() {
+      if (!this.isCollapsed) return;
+      contentEl.classList.remove('collapsed');
 
-    // let CSS transition max height from whatever it was to 0
-    contentEl.style.maxHeight = '';
-    contentEl.dispatchEvent(new Event('readmore-collapse'));
+      // this makes the CSS transition work, gracefully scaling from 100% to 0
+      contentEl.style.maxHeight = `${contentEl.scrollHeight}px`;
+      contentEl.style.webkitMaskImage =
+        contentEl.style.maskImage =
+          `linear-gradient(to bottom, white 100%, transparent)`;
+      contentEl.dispatchEvent(new Event('readmore-expand'));
+    },
+    get isCollapsed() {
+      return contentEl.classList.contains('collapsed');
+    },
+    get isOverflowed() {
```

Comment:
since calling this getter causes the dom to reflow and can potentially cause layout thrashing, you should make it more explicitly a function:
https://stackoverflow.com/questions/27637184/what-is-dom-reflow

The implication of a getter is that they're essentially free to call, and so a developer will never feel the need to store a local copy of a call to a getter.

Reply:
Wow, I actually didn't know that measuring an element can cause a reflow. Thanks for the tip